provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs

```diff
@@ -72,62 +72,77 @@ impl IterableResourceProvider<LikelySubtagsV1Marker> for LikelySubtagsProvider {
+            // Errors if the delta is not assignable to the pattern.
+            macro_rules! with_diff {
+                ($pat:pat => $stmt:expr ) => {
+                    if let $pat = (
+                        if entry.0.language != entry.1.language {
+                            entry.1.language
+                        } else {
+                            Language::UND
+                        },
+                        if entry.0.script != entry.1.script {
+                            entry.1.script
+                        } else {
+                            None
+                        },
+                        if entry.0.region != entry.1.region {
+                            entry.1.region
+                        } else {
+                            None
+                        },
+                    ) {
+                        $stmt;
+                    } else {
+                        panic!(
+                            "The expansion {:?} -> {:?} can not be stored in the pattern {}",
+                            entry.0,
+                            entry.1,
+                            stringify!($pat)
+                        );
+                    }
+                };
+            }
+
             if !entry.0.language.is_empty() {
-                let lang = entry.0.language.into();
+                let lang = entry.0.language;
                 if let Some(script) = entry.0.script {
-                    language_script.insert((lang, script.into()), extract_result(entry));
+                    with_diff!((Language::UND, None, Some(region)) => language_script.insert(&(lang, script), &region));
                 } else if let Some(region) = entry.0.region {
-                    language_region.insert((lang, region.into()), extract_result(entry));
+                    with_diff!((Language::UND, Some(script), None) => language_region.insert(&(lang, region), &script));
                 } else {
-                    language.insert(lang, extract_result(entry));
+                    with_diff!((Language::UND, Some(script), Some(region)) => language.insert(&lang, &(script, region)));
                 }
             } else if let Some(scr) = entry.0.script {
-                if let Some(reg) = entry.0.region {
-                    script_region.insert((scr.into(), reg.into()), extract_result(entry));
+                if let Some(region) = entry.0.region {
+                    with_diff!((language, None, None) => script_region.insert(&(scr, region), &language));
                 } else {
-                    script.insert(scr.into(), extract_result(entry));
+                    with_diff!((language, None, Some(region)) => script.insert(&scr, &(language, region)));
                 }
             } else if let Some(reg) = entry.0.region {
-                region.insert(reg.into(), extract_result(entry));
+                // Some of the target regions here are not equal to the source, such as und-002 -> en-Latn-NG.
```

Comment:
Calling this bit out for @zbraniecki or @sffc

Reply:
For for those entries, like "und-002 -> en-Latn-NG", does the current code have the same behavior as the new code?

Reply:
Yes, a subtag only gets updated if it's non-empty: https://github.com/unicode-org/icu4x/blob/main/components/locale_canonicalizer/src/locale_canonicalizer.rs#L202

I'm a bit sceptical about the correctness of that though, if CLDR gives us this data, maybe the reference algorithm doesn't actually discard it.

Reply:
I'm ok with this for now. We may want to dwell into this detail if we hit a case where it matters and it shouldn't be hard to fine-tune the code.